New import translator: Citavi 5 XML.js

zuphilip opened the pull request:
This is just a first try for an import translator for the Citavi 5 XML format, which the user can have when the make a backup save in Citavi (and unzip this `ctv5bak` file).

The PR is to have some place to discuss.

adam3smith commented:
Let me know when would be a good time for me to take a look.

zuphilip commented:
At the moment I have some problems to nest the different collections inside each other (see code which is commented out). Do you know how the `complete()` on collection in this case exactly work?

The documentation https://www.zotero.org/support/dev/translators/coding#creating_collections reads
> The children of a collection can include other collections. In this case, `collection.complete()` should be called only on the top-level collection. 

I tried different things but nothing worked really with nesting collections :-(

adam3smith commented:
I've never worked with that either, unfortunately, but I'm happy to try some stuff out. Could you upload a test citavi-xml file somewhere?

zuphilip commented:
Sure, the large examples with nested collections is here https://gist.github.com/zuphilip/76ce89ebbdac0386507b36cff3fd499a (besides there are some smaller examples in the test cases).

zuphilip commented on Citavi 5 XML.js:
```diff
@@ -0,0 +1,776 @@
+			Z.debug(collection.id);
+			Z.debug(subcategories);
+			for (var j=1; j<subcategories.length; j++) {
+				collection.children.push({ type: 'collection', id: subcategories[j] });
```
My impression is that everything here works, but the `complete()`-statements are not clear for me...

zuphilip replied:
This is form the code which is commented out.

zuphilip commented:
I added some more code and also added numbers to the categories indicating the hierarchy. This should be a good start for this translator which is already quite complicated...

zuphilip commented:
Any time now would be fine for reviewing the translator.

adam3smith commented:
Are collections working now, or should I test them still?

Sent from my phone

On Dec 26, 2016 11:46 PM, "Philipp Zumstein" <notifications@github.com>
wrote:

Any time now would be fine for reviewing the translator.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/zotero/translators/pull/1189#issuecomment-269246863>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAH32ssiDsF96uIzt0z17_EYHukK9OLBks5rMEPKgaJpZM4K85Ty>
.

zuphilip commented:
(I tried really quite some time, but w/o any idea how this might work it seems for me not the best option to spend time.)

Different collections work but nesting (collections in collections) is not working. However, I added the numbering also to the collection names and thereby the information is kind of preserved:
![outliner](https://cloud.githubusercontent.com/assets/5199995/21500346/73cf6690-cc3e-11e6-8363-7f8d2932767f.png)

There is no outlining feature in Zotero's word processor in contrast to Citavi's word processor. Thus, I guess it is not the same important.

adam3smith commented on Citavi 5 XML.js:
```diff
@@ -0,0 +1,980 @@
+			Z.debug("Therefore use default type 'journalArticle'");
+			item = new Zotero.Item("journalArticle");
+		}
+		item.itemID = ZU.xpathText(references[i], './@id');
```
what's the point of setting an item ID here? Zotero doesn't do anything with those. (I'll post these as single comments, since I'm not sure how quickly I'll be able to get through all of them).

zuphilip replied:
They are needed for the `seeAlso`. (Moreover, I used them internally for several things.)

adam3smith commented on Citavi 5 XML.js:
```diff
@@ -0,0 +1,980 @@
+			var addressType = ZU.xpathText(locations[j], 'MirrorsReferencePropertyId');
+			if (address) {
+				if (addressType == "Doi" && !item.DOI) {
+					item.DOI = address;
```
how about adding the DOI to Extra for items that aren't journal articles or conference papers? (same above)

zuphilip replied:
Yes, that is a good idea. I will do that.

What exactly do you mean by "(same above)"?

adam3smith replied:
I just meant to add the same logic around line 136 where you add the regular `item.DOI` (or just add it once here, but make sure to also cover the case where `item.DOI` already exists.)

adam3smith commented on Citavi 5 XML.js:
```diff
@@ -0,0 +1,980 @@
+			var libraryId = ZU.xpathText(locations[j], 'LibraryID');
+			if (callNumber) {
+				item.callNumber = callNumber;
+				item.libraryCatalog = ZU.xpathText(doc.getElementById(libraryId), "Name");
```
this seems too restrictive to me. Zotero saves a library Catalog for every item imported from the web. We should probably always import this? (I'm saying this not knowing Citavi)

zuphilip replied:
In Citavi you can have several libraries and for each item it can be saved that it is present in a library and additionally you can also add a call number then. However, you can also have libraries associated to an item without a call number, e.g.

![libraries](https://cloud.githubusercontent.com/assets/5199995/21534635/5bb16b76-cd6a-11e6-8734-01494684d57b.png)

On the other hand Zotero cannot handle multiplies libraryCatalogues...

zuphilip replied:
By deleting the condition as you suggested, it may (depending on the order the objects are processed) lead to the situation that the call number does not belong to the library saved in Zotero.

adam3smith replied:
you know the data model better but could it then make sense to do

```
else {
  item.libraryCatalog = ZU.xpathText(doc.getElementById(libraryId), "Name");
}
```
or some other way that we can make sure that we can make sure that 
1) if an item has both a library catalog and a call number they belong to the same catalog and
2) if an item only has a library catalog, it still gets added to the item

zuphilip replied:
Well, this works for single library catalogues with and without call numbers. But here we are in a `for`-loop which can cause problems. Let us assume in the example above we first have the case "Hochschule Hannover" with call number "NRT 2950/5" which will be saved in Zotero in the `libraryCatalog` and `callNumber`. Afterwards we process the case "TIB/UB Hannover" and will replace (only!) the `libraryCatalog`. That means that afterwards we would have saved "TIB/UB Hannover" and "NRT 2950/5" in Zotero which will not match to each other.

However, I can save the information in a temporary array first and then afterwards do a better assigning to the `item`. I will give it a try ... (and hope it is not getting to complicated code...)

adam3smith commented on Citavi 5 XML.js:
```diff
@@ -0,0 +1,980 @@
+		item.complete();
+	}
+	
+	//Categories will be mapped to collections where the name contains
```
note to self: reviewed until here. 2016/12/28

adam3smith commented:
OK, apart from the DOI issue we agree on and the open question on Library Catalog, this looks good to me.

zuphilip commented:
New version:
 * Copy the DOI into an extra line for itemTypes not supporting (yet) this field
 * Handle cases when callNumber or libraryCatalog is missing.
 * I had to store the test data also outside, because Scaffold did delete that every time.

Please have a look and let me know what you think.

adam3smith commented:
Too bad about the tests, but I don't think we should delay because of this -- any idea what's going on, though?
Otherwise this looks good to me, but I've only reviewed code, haven't tested. If you think it's ready, I'll merge.

zuphilip commented:
> Too bad about the tests, but I don't think we should delay because of this -- any idea what's going on, though?

Not really. It is for me at the moment like "Stochern-im-Dunkeln". It could be connected to some encoding between Windows, Linux during Git commands, but maybe it is something different...

> Otherwise this looks good to me, but I've only reviewed code, haven't tested. If you think it's ready, I'll merge.

From my side this is okay and ready. There might be improvements or nits occurring when more people are trying it out, but I don't see anything which could be blocking at the moment.

(Must new import formats/translators be discussed with the core Zotero developers?)

adam3smith commented:
> (Must new import formats/translators be discussed with the core Zotero developers?)

I don't think so. @dstillman has an eye on this repository, and will comment if he's unhappy with something, but otherwise I only check with him on potentially controversial things (e.g. for export formats like CSV that don't roundtrip)

Thanks for this -- it's terriffic.

zuphilip commented:
🎉 🎁 🎆
